### ISSUE TITLE
How to get signature from PublicKey + Timestamp bytes

### ISSUE THREAD

**Issue by wrwc** [wavesplatform/Waves#2625]
where signature was generated from PublicKey + Timestamp bytes ?

**Comment by username1565** [comment]
[Node REST API](https://nodes.wavesplatform.com/) allow this
See:
```
POST /addresses/signText/{address}
POST /addresses/sign/{address}
POST /addresses/verifyText/{address}
POST /addresses/verify/{address}
```
Address must be a node address - the address, which already imported on the node.

Also, you can sign the message by privkey:
`
POST /utils/sign/{privateKey}
`
Will be better to add signing and verify functions inside waves-client, [because there](https://client.wavesplatform.com/#!/dex-demo) is possible to log-in by privkey.

**Comment by wrwc** [comment]
@username1565 ,

Thankyou !!!

But still now i can't get the order-book ( matcher ) response .

I am getting signature & message value form testnet ( https://testnodes.wavesnodes.com/api-docs/index.html ) using privatekey.

PFA.


![Swagger UI](https://user-images.githubusercontent.com/55442133/65580040-8d497200-df96-11e9-8d8a-604b960324ab.png)
![Screenshot from 2019 09 25 13 10 10](https://user-images.githubusercontent.com/55442133/65580046-90446280-df96-11e9-9ee0-0314f731239c.png)

Kindly provide the solution ASAP.

Please Help Me.

**Comment by vsuharnikov** [comment]
Hello! 

We haven't such REST API method. IMO it's unsafe to send your private key to the NODE and it's better to sign the request by yourself.

## A simple way

Use one of the client libraries: https://docs.wavesplatform.com/en/waves-api-and-sdk/client-libraries.html

## You-really-should-think-twice-or-more-times-way

A signature is required for methods:
* `GET /orderbook/{amountAsset}/{priceAsset}/publicKey/{publicKey}` - an order history by an asset pair and a public key (**1**);
* `GET /orderbook/{publicKey}` - an order history by a public key (**1**);
* `GET /balance/reserved/{publicKey}` - a reserved balance by a public key (**1**);
* `POST /matcher/orderbook/{amountAsset}/{priceAsset}/cancel` - cancel either a single order (**2**) or all orders by pair (**3**);
* `POST /matcher/orderbook/cancel` - cancel all orders (**3**);

Signatures:
1. You should provide two HTTP headers: *Timestamp* (in milliseconds) and *Signature* (in a base 58 representation)

    `Signature  = encodeAsBase58String(sign(privateKey, concat(publicKey, timestampInMillisAsLong)))`

2. You should provide a JSON in a body of request:

    ```json
    {
      "sender": "your public key in a base58 encoded string",
      "orderId": "order id in a base58 encoded string",
      "signature": "signature in a base58 encoded string"
    }
    ```

    `signature = encodeAsBase58String(sign(privateKey, concat(publicKey, orderId)))`

2. You should provide a JSON in a body of request:

    ```json
    {
      "sender": "your public key in a base58 encoded string",
      "timestamp": timestampInMillis,
      "signature": "signature in a base58 encoded string"
    }
    ```

    `signature = encodeAsBase58String(sign(privateKey, concat(publicKey, timestampInMillis)))`

**Comment by wrwc** [comment]
@vsuharnikov 

Hi , 

Kindly check my code below ,

const { base58Decode, base58Encode,concat, publicKey, signBytes } = require('@waves/ts-lib-crypto');

let timestamp = +(new Date().getTime());

let signatures = base58Encode(signBytes('GfSvHejBDJW77avjPvn9ZEsA6KA5ecF8eJWTFgd7wJgQ', concat('3sAZVZZwhP9sGG6xUA3snihb1T9HmvsyfrW5GbvRNiD2', timestamp)))

Response : 

signature :  4aJKPVHVYkCDHTfAiytKEdGwKM57KtVdU35ndehwtvCNkZxUJcsFHwqa9N9kDBgQS65F5UmWTrPaotTAM4jeybWY

But still now i am facing same error 

{
  "result": null,
  "template": "The request has an invalid signature",
  "success": false,
  "error": 1051904,
  "message": "The request has an invalid signature",
  "status": "InvalidSignature"
}

Can you please check and let me know what we wrong.

Please Help me

**Comment by username1565** [comment]
Hello, @vsuharnikov! I have another questions about signing.
I see libraries, link on which you privided here, working with transactions.
But... What about singing and verify messages?
I think this is possible to make client-side, so I did open [this issue as future request for WAVES-client](https://github.com/wavesplatform/WavesGUI/issues/1610).

**TL;DR**, go to **Summary**.
But... I see some methods in the [NODE REST API](https://nodes.wavesplatform.com/).
For example:
```
POST /utils/sign/{privateKey} 
```
Can signing the messages (without spaces, because error).

Example of using:
1. https://nodes.wavesplatform.com/
2. POST /utils/sign/{privateKey}
privateKey: "3kMEhU5z3v8bmer1ERFUUhW58Dtuhyo9hE5vrhjqAWYT" (default private key)
message: "testmessagetextwithoutspaces"
3. Click "Try it out!" button.
4. Response Body:
```
{
  "message": "testmessagetextwithoutspaces",
  "signature": "24s5ksZVFVQujdBfVE6vUFmaJCq1mN8WG6Qtwqzq1LhT3dHETyVHtzEErEkXhGFpTXi3U1KzaD1EPe7KV7qv635z"
}
```
There is errors, if spaces will be inside the message:
Example:
message: "test_message text-without+spaces"
Response Body:
```
{
  "error": 0,
  "message": "Error is unknown"
}
```
Can you fix this there?


As you can see, this method can signing the message, self-valid,
because there is no privided the one parameter:
```
"publickey": "string"
```
Can you add this publickey, there, and fix signing for that method?

Next thing...
I did try to log-in to https://client.wavesplatform.com/#!/dex-demo
using this privkey.
In the settings, I see for privkey: 3kMEhU5z3v8bmer1ERFUUhW58Dtuhyo9hE5vrhjqAWYT
is corresponding the address: 3PPbMwqLtwBGcJrTA5whqJfY95GqnNnFMDX
and pubkey: HBqhfdFASRQ5eBBpu2y6c6KKi1az6bMx8v1JxX4iW1Q8

So, I did modify response, up to:
```
{
  "message": "testmessagetextwithoutspaces",
  "signature": "24s5ksZVFVQujdBfVE6vUFmaJCq1mN8WG6Qtwqzq1LhT3dHETyVHtzEErEkXhGFpTXi3U1KzaD1EPe7KV7qv635z",
  "publickey": "HBqhfdFASRQ5eBBpu2y6c6KKi1az6bMx8v1JxX4iW1Q8"
}
```

Then, I did try to verify this signed message, using methods:
```
POST /addresses/verifyText/{address} 
```
and
```
POST /addresses/verify/{address} 
```
But... This need to provide API_KEY on my own node.
```
{
  "error": 2,
  "message": "Provided API key is not correct"
}
```
So I did using this API-KEY, on my node, and then, after explore this with my API-key,
I did go to:
```
POST /addresses/verifyText/{address}
```
and put there the following data:
1. address: 3PPbMwqLtwBGcJrTA5whqJfY95GqnNnFMDX
2. body:
```
{
  "message": "testmessagetextwithoutspaces",
  "signature": "24s5ksZVFVQujdBfVE6vUFmaJCq1mN8WG6Qtwqzq1LhT3dHETyVHtzEErEkXhGFpTXi3U1KzaD1EPe7KV7qv635z",
  "publickey": "HBqhfdFASRQ5eBBpu2y6c6KKi1az6bMx8v1JxX4iW1Q8"
}
```
3. Click "Try it out!" button...
4. Response Body
```
{
  "valid": false
}
```

Then, I did try to verify this, using directive:
```
POST /addresses/verify/{address}
```
1. address: 3PPbMwqLtwBGcJrTA5whqJfY95GqnNnFMDX
2. body:
```
{
  "message": "testmessagetextwithoutspaces",
  "signature": "24s5ksZVFVQujdBfVE6vUFmaJCq1mN8WG6Qtwqzq1LhT3dHETyVHtzEErEkXhGFpTXi3U1KzaD1EPe7KV7qv635z",
  "publickey": "HBqhfdFASRQ5eBBpu2y6c6KKi1az6bMx8v1JxX4iW1Q8"
}
```
3. Click "Try it out!" button...
4. Response Body
```
{
  "valid": true
}
```
_______________________________________
**Summary:**
1. `POST /utils/sign/{privateKey}` is working self-correct, because "publickey" not specified in the response, and when message contains spaces - this not working correctly. Need to fix.
2. Do verify this can only the directive `POST /addresses/verify/{address}`, if correct public-key specified on private node.
So, can you add some method on the node, to get public key from private key?
3. Is this possible to do signing and verify messages client-side? Is there some scritps?

That's all. Have a nice day.

**Comment by username1565** [comment]
@wrwc, maybe, this code can help you:
https://waveswallet.io/js/waves-lite-client-mainnet-0.5.22.js
```
        function buildLoadUserOrdersSignature(timestamp, sender) {
            validateService.validateSender(sender);

            var publicKeyBytes = utilityService.base58StringToByteArray(sender.publicKey),
                timestampBytes = utilityService.longToByteArray(timestamp),
                signatureData = [].concat(publicKeyBytes, timestampBytes),

                privateKeyBytes = cryptoService.base58.decode(sender.privateKey);

            return cryptoService.nonDeterministicSign(privateKeyBytes, signatureData);
        }
```

Also, you can save this script, as `.js`-file, then create `index.html` with code:
```
<script src="waves-lite-client-mainnet-0.5.22.js"></script>
```
open this `index.html` in browser, and in console (F12 button), run the command:
```
axlsign.signMessage(
  new Uint8Array([privkey_bytes]), 
  new Uint8Array([msg_bytes])
);//return Uint8Array
```
and the command:
```
axlsign.verify(
	new Uint8Array([publicKeyBytes]),
	new Uint8Array([msgBytes]),
	new Uint8Array([signatureBytes])
) //Return true (verified), and false (if invalid)
```
Example, the time-stamped message, signed by my private key of my node:
```
axlsign.verify(
	new Uint8Array([17, 90, 98, 207, 48, 199, 160, 241, 21, 151, 155, 191, 107, 73, 192, 163, 37, 35, 33, 66, 67, 223, 220, 131, 108, 84, 238, 211, 87, 103, 18, 74]),
	new Uint8Array([17, 90, 98, 207, 48, 199, 160, 241, 21, 151, 155, 191, 107, 73, 192, 163, 37, 35, 33, 66, 67, 223, 220, 131, 108, 84, 238, 211, 87, 103, 18, 74, 0, 0, 1, 109, 142, 54, 112, 176]),
	new Uint8Array([160, 151, 13, 102, 84, 85, 76, 167, 177, 178, 29, 175, 118, 18, 112, 69, 242, 224, 236, 63, 19, 25, 104, 143, 205, 109, 213, 198, 197, 141, 107, 105, 232, 0, 141, 33, 52, 122, 3, 55, 218, 52, 206, 81, 21, 3, 222, 122, 32, 157, 190, 180, 80, 166, 29, 76, 8, 30, 19, 189, 186, 239, 91, 133])
) //true
```
As you can see, the `msgBytes`-array contains the part of `publicKeyBytes` + `timestampBytes`.

Maybe, this is the best way to make sign and verify messages - `client-side`.
But I don't know how to verify this signature on the node.

The both directives
```
POST /addresses/verify/{address}
```
and
```
POST /addresses/verifyText/{address}
```
not working with bytes, or base64 encoded messages:
```
{
  "error": 110,
  "message": "invalid message"
}
```
But... As you can see, signature was been verified sucessfully, `client-side`, in previous example.
So, see the source code of this scripts, to find more explanation.
Also, you can see the source code of stable version old waves-lite-client v0.5.18a,
here: https://github.com/username1565/waves-lite-client
